File: frontend/dev.js
#!/usr/bin/env node

/**
 * Frontend development helper script
 */

const { spawn } = require('child_process');
const fs = require('fs');
const path = require('path');

// Helper function to run commands
function runCommand(command, args = [], options = {}) {
  return new Promise((resolve, reject) => {
    console.log(`🔧 Running: ${command} ${args.join(' ')}`);
    
    const child = spawn(command, args, {
      stdio: 'inherit',
      shell: true,
      ...options
    });

    child.on('close', (code) => {
      if (code === 0) {
        console.log(`✅ Command completed successfully`);
        resolve(code);
      } else {
        console.log(`❌ Command failed with code ${code}`);
        reject(new Error(`Command failed with code ${code}`));
      }
    });

    child.on('error', (error) => {
      console.log(`❌ Command error: ${error.message}`);
      reject(error);
    });
  });
}

// Check if file exists
function fileExists(filePath) {
  return fs.existsSync(filePath);
}

// Setup development environment
async function setup() {
  console.log('🚀 Setting up frontend development environment...');

  // Check if package.json exists
  if (!fileExists('package.json')) {
    console.log('❌ package.json not found. Are you in the frontend directory?');
    process.exit(1);
  }

  // Check if .env.local exists
  if (!fileExists('.env.local')) {
    if (fileExists('.env.local.example')) {
      fs.copyFileSync('.env.local.example', '.env.local');
      console.log('📝 Created .env.local from .env.local.example');
      console.log('⚠️  Please edit .env.local with your configuration');
    } else {
      console.log('⚠️  No .env.local.example found');
    }
  }

  // Install dependencies
  try {
    await runCommand('npm', ['install']);
  } catch (error) {
    console.log('❌ Failed to install dependencies');
    process.exit(1);
  }

  console.log('✅ Frontend development environment setup complete!');
  console.log('\nNext steps:');
  console.log('1. Edit .env.local with your backend URL');
  console.log('2. Run: node dev.js serve');
}

// Start development server
async function serve() {
  console.log('🚀 Starting development server...');
  try {
    await runCommand('npm', ['run', 'dev']);
  } catch (error) {
    console.log('❌ Failed to start development server');
    process.exit(1);
  }
}

// Build for production
async function build() {
  console.log('🏗️  Building for production...');
  try {
    await runCommand('npm', ['run', 'build']);
    console.log('✅ Build completed successfully!');
  } catch (error) {
    console.log('❌ Build failed');
    process.exit(1);
  }
}

// Start production server
async function start() {
  console.log('🚀 Starting production server...');
  try {
    await runCommand('npm', ['run', 'start']);
  } catch (error) {
    console.log('❌ Failed to start production server');
    process.exit(1);
  }
}

// Lint code
async function lint() {
  console.log('🧹 Linting code...');
  try {
    await runCommand('npm', ['run', 'lint']);
    console.log('✅ Linting completed successfully!');
  } catch (error) {
    console.log('❌ Linting failed');
    process.exit(1);
  }
}

// Format code
async function format() {
  console.log('🎨 Formatting code...');
  try {
    await runCommand('npm', ['run', 'format']);
    console.log('✅ Code formatting completed!');
  } catch (error) {
    console.log('❌ Code formatting failed');
    process.exit(1);
  }
}

// Type check
async function typeCheck() {
  console.log('🔍 Type checking...');
  try {
    await runCommand('npm', ['run', 'type-check']);
    console.log('✅ Type checking completed!');
  } catch (error) {
    console.log('❌ Type checking failed');
    process.exit(1);
  }
}

// Clean build artifacts
async function clean() {
  console.log('🧹 Cleaning build artifacts...');
  try {
    await runCommand('npm', ['run', 'clean']);
    console.log('✅ Cleanup completed!');
  } catch (error) {
    console.log('❌ Cleanup failed');
    process.exit(1);
  }
}

// Install new dependency
async function install(packageName) {
  if (!packageName) {
    console.log('❌ Please provide a package name');
    console.log('Usage: node dev.js install <package-name>');
    process.exit(1);
  }

  console.log(`📦 Installing ${packageName}...`);
  try {
    await runCommand('npm', ['install', packageName]);
    console.log(`✅ ${packageName} installed successfully!`);
  } catch (error) {
    console.log(`❌ Failed to install ${packageName}`);
    process.exit(1);
  }
}

// Show help
function showHelp() {
  console.log(`
🛠️  Frontend Development Helper

Usage: node dev.js <command> [options]

Commands:
  setup       Set up development environment
  serve       Start development server
  build       Build for production
  start       Start production server
  lint        Lint code
  format      Format code
  typecheck   Run type checking
  clean       Clean build artifacts
  install     Install new dependency
  help        Show this help message

Examples:
  node dev.js setup              # First time setup
  node dev.js serve              # Start development server
  node dev.js build              # Build for production
  node dev.js install axios      # Install new dependency
  node dev.js lint               # Lint all code
`);
}

// Main function
async function main() {
  const command = process.argv[2];
  const args = process.argv.slice(3);

  switch (command) {
    case 'setup':
      await setup();
      break;
    case 'serve':
      await serve();
      break;
    case 'build':
      await build();
      break;
    case 'start':
      await start();
      break;
    case 'lint':
      await lint();
      break;
    case 'format':
      await format();
      break;
    case 'typecheck':
      await typeCheck();
      break;
    case 'clean':
      await clean();
      break;
    case 'install':
      await install(args[0]);
      break;
    case 'help':
    default:
      showHelp();
      break;
  }
}

main().catch((error) => {
  console.error('❌ An error occurred:', error.message);
  process.exit(1);
});

File: backend/dev.py
#!/usr/bin/env python3
"""
Development helper script for the Text Summarization API.
"""

import os
import sys
import subprocess
from pathlib import Path


def run_command(command, description):
    """Run a shell command with description."""
    print(f"🔧 {description}...")
    result = subprocess.run(command, shell=True, capture_output=True, text=True)
    if result.returncode == 0:
        print(f"✅ {description} completed successfully")
        if result.stdout:
            print(result.stdout)
    else:
        print(f"❌ {description} failed")
        if result.stderr:
            print(result.stderr)
        return False
    return True


def install_dependencies():
    """Install Python dependencies."""
    return run_command("pip install -r requirements.txt", "Installing dependencies")


def run_tests():
    """Run the test suite."""
    return run_command("pytest tests_new/ -v", "Running tests")


def run_server():
    """Run the development server."""
    print("🚀 Starting development server...")
    subprocess.run("uvicorn src.main:app --reload --host 0.0.0.0 --port 8000", shell=True)


def format_code():
    """Format code with black and isort."""
    commands = [
        ("black src/ tests_new/", "Formatting code with black"),
        ("isort src/ tests_new/", "Sorting imports with isort")
    ]
    
    for command, description in commands:
        if not run_command(command, description):
            return False
    return True


def lint_code():
    """Lint code with flake8."""
    return run_command("flake8 src/ tests_new/", "Linting code")


def type_check():
    """Run type checking with mypy."""
    return run_command("mypy src/", "Type checking")


def build_docker():
    """Build Docker image."""
    return run_command("docker build -t text-summarization-api .", "Building Docker image")


def run_docker():
    """Run Docker container."""
    print("🐳 Starting Docker container...")
    subprocess.run(
        "docker run -p 8000:8000 --env-file .env text-summarization-api", 
        shell=True
    )


def setup_dev_environment():
    """Set up development environment."""
    print("🔧 Setting up development environment...")
    
    # Check if .env exists
    if not Path(".env").exists():
        if Path(".env.example").exists():
            print("📝 Creating .env from .env.example...")
            subprocess.run("cp .env.example .env", shell=True)
            print("⚠️  Please edit .env with your configuration")
        else:
            print("❌ No .env.example found")
            return False
    
    # Install dependencies
    if not install_dependencies():
        return False
    
    # Create uploads directory
    Path("uploads").mkdir(exist_ok=True)
    print("📁 Created uploads directory")
    
    print("✅ Development environment setup complete!")
    print("\nNext steps:")
    print("1. Edit .env with your configuration (especially GROQ_API_KEY)")
    print("2. Run: python dev.py serve")
    
    return True


def show_help():
    """Show help message."""
    print("""
🛠️  Text Summarization API Development Helper

Usage: python dev.py <command>

Commands:
  setup       Set up development environment
  install     Install dependencies
  serve       Run development server
  test        Run tests
  format      Format code (black + isort)
  lint        Lint code (flake8)
  typecheck   Run type checking (mypy)
  docker      Build and run Docker container
  build       Build Docker image only
  help        Show this help message

Examples:
  python dev.py setup      # First time setup
  python dev.py serve      # Start development server
  python dev.py test       # Run all tests
  python dev.py format     # Format all code
""")


def main():
    """Main entry point."""
    if len(sys.argv) != 2:
        show_help()
        return
    
    command = sys.argv[1].lower()
    
    commands = {
        'setup': setup_dev_environment,
        'install': install_dependencies,
        'serve': run_server,
        'test': run_tests,
        'format': format_code,
        'lint': lint_code,
        'typecheck': type_check,
        'build': build_docker,
        'docker': lambda: build_docker() and run_docker(),
        'help': show_help
    }
    
    if command in commands:
        commands[command]()
    else:
        print(f"❌ Unknown command: {command}")
        show_help()


if __name__ == "__main__":
    main()
